File: functions/CopBET_von_Neumann_entropy.m
% out = CopBET_von_Neumann_:entropy(in,keepdata)
%
% Copenhagen Brain Entropy Toolbox: Von-Neumann entropy.
% This function evaluates Von-Neumann entropy as introduced in Felippe et
% al., 2021. The Von-Neumann entropy is a Shannon entropy of a correlation
% matrix scaled by the number of regions. 
%
% Input:
%   in: a matrix (nxp,n>1) or a table where the first column contains
%   matrices (in cells) to be concatenated before clustering, e.g.,
%   different subjects or scan sessions.
% name-value pairs:
%   keepdata: Indicates whether the output table also should contain the
%   input data, i.e., by adding an extra column containing entropy values.
%   Defaults to true
%
% Neurobiology Research Unit, 2023
% Please cite McCulloch, Olsen et al., 2023: "Navigating Chaos in
% Psychedelic Neuroimaging: A Rigorous Empirical Evaluation of the Entropic
% Brain Hypothesis" if you use CopBET in your studies. Please read the
% paper to get a notion of our recommendations regarding the use of the
% specific methodologies in the toolbox.

% ASO 9/3-2023

% Copyright (C) 2023 Anders Stevnhoved Olsen & Drummond E-Wen McCulloch
%
%    This program is free software: you can redistribute it and/or modify
%    it under the terms of the GNU General Public License as published by
%    the Free Software Foundation, either version 3 of the License, or
%    (at your option) any later version.
%
%    This program is distributed in the hope that it will be useful,
%    but WITHOUT ANY WARRANTY; without even the implied warranty of
%    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%    GNU General Public License for more details.
%
%    You should have received a copy of the GNU General Public License
%    along with this program.  If not, see http://www.gnu.org/licenses/.

function out = CopBET_von_Neumann_entropy(in,varargin)
[out,numworkers,in] = CopBET_function_init(in,varargin);

%load data
entropy = nan(height(in),1);
for ses = 1:height(in)
    
    R = corrcoef(in{ses,1}{1});
    
    rho = R./size(R,1);
    
%     entropy(ses) = -trace(rho*logm(rho));
    
    eigvals = eig(rho);
    
    % sometimes the eigenvalues are machine-precision negative...
    eigvals(eigvals<0)=eps;
    entropy(ses) = -sum(eigvals.*log(eigvals));
    
    % Normalize by maximum entropy
    entropy(ses) = entropy(ses)./log(size(R,1));
    
end

out.entropy = entropy;
end

File: copbet_py/functions/CopBET_von_Neumann_entropy.py
"""
CopBET_von_Neumann_entropy
==========================
Copenhagen Brain Entropy Toolbox: Von-Neumann entropy.

Evaluates Von-Neumann entropy as introduced in Felippe et al., 2021.
The Von-Neumann entropy is computed from the eigenvalues of the
density matrix rho = R / N, where R is the Pearson correlation matrix
and N is the number of regions.

Input
-----
sessions : list of np.ndarray, each shape (T, N)
    Time series for each session (T timepoints, N regions).

Returns
-------
entropy : np.ndarray, shape (n_sessions,)
    Normalized Von-Neumann entropy per session.

Reference
---------
Felippe et al., 2021.

Please cite McCulloch, Olsen et al., 2023 if you use CopBET.
"""

import numpy as np


def CopBET_von_Neumann_entropy(sessions):
    """
    Compute Von-Neumann entropy for a list of fMRI sessions.

    Parameters
    ----------
    sessions : list of np.ndarray, each shape (T, N)

    Returns
    -------
    entropy : np.ndarray, shape (n_sessions,)
    """
    entropy = np.full(len(sessions), np.nan)

    for ses, ts in enumerate(sessions):
        print(f"Processing session {ses + 1}/{len(sessions)}...", end='\r')
        ts = np.asarray(ts, dtype=float)
        N = ts.shape[1]

        R = np.corrcoef(ts.T)  # (N, N) correlation matrix

        # Density matrix: rho = R / N
        rho = R / N

        # Eigenvalues of rho
        eigvals = np.linalg.eigvalsh(rho)

        # # Clamp tiny negatives from floating point
        # eigvals = np.where(eigvals < 0, np.finfo(float).eps, eigvals)

        # Von-Neumann entropy: S = -Tr(rho * log(rho)) = -sum(lambda * log(lambda))
        # Avoid log(0) by masking near-zero eigenvalues
        mask = eigvals > 0
        H = -np.sum(eigvals[mask] * np.log(eigvals[mask]))

        # Normalize by maximum entropy log(N)
        entropy[ses] = H / np.log(N)

    return entropy
